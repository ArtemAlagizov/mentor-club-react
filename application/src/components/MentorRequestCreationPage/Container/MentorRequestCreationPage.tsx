import React, { ChangeEvent, useEffect, useState } from "react";
import { connect } from "react-redux";
import MentorRequestCreationPageView from "../View/MentorRequestCreationPageView";
import {
  clearNewMentorRequestId,
  createMentorRequestBegin,
  fetchMentorListBegin
} from "../../../state/MentorRequests/actions";
import { IAppState } from "../../../store";
import {
  selectIsFetchingMentors,
  selectMentors,
  selectNewMentorRequestId
} from "../../../state/MentorRequests/selectors";
import { IMentorRequestCreationPageProps } from "./IMentorRequestCreationPageProps";
import { useHistory } from "react-router";
import { IUser } from "../../../state/MentorRequests/types";

export const MentorRequestCreationPage: React.FC<IMentorRequestCreationPageProps> = ({
  mentors,
  fetchMentorListBegin,
  createMentorRequestBegin,
  clearNewMentorRequestId,
  isFetchingMentorList,
  newMentorRequestId
}) => {
  const history = useHistory();
  const [selectedMentor, setSelectedMentor] = useState<IUser | null>(null);
  const [mentorRequestDescription, setMentorRequestDescription] = useState<
    string
  >("");

  useEffect(() => {
    fetchMentorListBegin();
  }, [fetchMentorListBegin]);

  useEffect(() => {
    if (newMentorRequestId) {
      history.push(`/mentor-requests/${newMentorRequestId}`);
    }
  }, [newMentorRequestId, history]);

  useEffect(() => {
    return () => {
      clearNewMentorRequestId();
    };
  }, [clearNewMentorRequestId]);

  const updateMentorRequestDescription = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => setMentorRequestDescription(event.target.value);

  const createMentorRequest = () => {
    createMentorRequestBegin({
      mentorRequestId: "",
      requestDescription: "",
      requesterName: "",
      status: "CREATED",
      requesterPictureThumbnailSrc: ""
    });
  };

  return (
    <MentorRequestCreationPageView
      mentors={mentors}
      selectedMentor={selectedMentor}
      setSelectedMentor={setSelectedMentor}
      mentorRequestDescription={mentorRequestDescription}
      updateMentorRequestDescription={updateMentorRequestDescription}
      isLoading={isFetchingMentorList}
      createMentorRequest={createMentorRequest}
    />
  );
};

export const mapStateToProps = (state: IAppState) => ({
  mentors: selectMentors(state),
  isFetchingMentorList: selectIsFetchingMentors(state),
  newMentorRequestId: selectNewMentorRequestId(state)
});

export const mapDispatchToProps = {
  fetchMentorListBegin,
  createMentorRequestBegin,
  clearNewMentorRequestId
} as const;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MentorRequestCreationPage);
