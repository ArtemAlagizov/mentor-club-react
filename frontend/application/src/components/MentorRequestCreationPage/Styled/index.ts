import styled from "styled-components";
import { SpinnerBoundaries } from "./SpinnerBoundaries";
import { Button } from "antd";

const Container = styled.div`
  padding: 15px;
  height: calc(100% - 45px);

  .ant-spin-nested-loading {
    height: 50px;
  }
`;

const NavigationButton = styled(Button)`
  border-radius: 0;
  bottom: 10px;
  right: 10px;
  min-width: 11rem;
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: 10px;
`;

const TopHeaderContainer = styled.div`
  display: flex;
  justify-content: space-between;
  color: white;
`;

const TopHeader = styled.div`
  font-size: 20px;
  color: white;
`;

const FormAndButtonContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
`;

const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

const CardContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  max-width: 400px;
  align-self: center;

  .ant-card-bordered {
    border: 0;
  }
`;

const UserImage = styled.img`
  margin-top: -3px;
  width: 3rem;
`;

export const Styled = {
  Container,
  TopHeaderContainer,
  TopHeader,
  SpinnerBoundaries,
  NavigationButton,
  ButtonContainer,
  FormAndButtonContainer,
  FormContainer,
  CardContainer,
  UserImage
};
