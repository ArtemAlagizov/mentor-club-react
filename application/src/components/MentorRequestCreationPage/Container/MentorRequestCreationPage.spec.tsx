import React from "react";
import { shallow } from "enzyme";
import { MentorRequestCreationPage } from "./MentorRequestCreationPage";

jest.mock("react-router-dom");
jest.mock("react-router", () => ({
  useHistory: () => ({
    push: jest.fn()
  })
}));

describe("Given a MentorRequestCreationPage", () => {
  it("should have MentorRequestResultPage matching snapshot", () => {
    const component = shallow(
      <MentorRequestCreationPage
        mentors={[]}
        isFetchingMentorList={false}
        clearNewMentorRequestId={jest.fn()}
        createMentorRequestBegin={jest.fn()}
        fetchMentorListBegin={jest.fn()}
        newMentorRequestId={""}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
