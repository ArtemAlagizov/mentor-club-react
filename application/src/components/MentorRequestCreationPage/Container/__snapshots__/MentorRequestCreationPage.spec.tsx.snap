// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MentorRequestCreationPage should have MentorRequestResultPage matching snapshot 1`] = `
<MentorRequestCreationPageView
  createMentorRequest={[Function]}
  isLoading={false}
  mentorRequestDescription=""
  mentors={Array []}
  selectedMentor={null}
  setSelectedMentor={[Function]}
  updateMentorRequestDescription={[Function]}
/>
`;
